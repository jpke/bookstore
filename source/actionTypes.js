export const SELECT_NEW_BOOK = 'SELECT_NEW_BOOK';
export const ADMIN_ADD_BOOK = 'ADMIN_ADD_BOOK';
export const USER_INFO = 'USER_INFO';
export const DELIVERY_AND_BILLING = 'DELIVERY_AND_BILLING';
export const DELIVERY = 'DELIVERY';
export const BILLING = 'BILLING';
export const PAYMENT_INFO = 'PAYMENT_INFO';
export const SUBMIT_ORDER = 'SUBMIT_ORDER';
