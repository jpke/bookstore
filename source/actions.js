import * as types from './actionTypes';

export function selectNewBook(newBook) {
  return {
    type: types.SELECT_NEW_BOOK,
    newBook
  };
}

function deliveryAddress(info, useDeliveryAddressAsBilling) {
  return {
    type: types.DELIVERY,
    info,
    useDeliveryAddressAsBilling
  };
}

function billingAddress(info) {
  return {
    type: types.BILLING,
    info,
  };
}

export function submitInfo(info, type, option) {
  if(type === 'AdminAddBook') {
    if(info.title === "" || info.author === "") return;
    return {
      type: types.ADMIN_ADD_BOOK,
      info
    };
  }
  else if(type === 'UserInfo') {
    return {
      type: types.USER_INFO,
      info
    };
  }
  else if(type === 'Delivery') {
    if(option) {
      //use delivery address as billing address
      return {
        type: types.DELIVERY_AND_BILLING,
        info,
        useDeliveryAddressAsBilling: option
      };
    }
    else {
      return {
        type: types.DELIVERY,
        info,
        useDeliveryAddressAsBilling: option
      };
    }
  }
  else if(type === 'Billing') {
    return {
      type: types.BILLING,
      info,
    };
  }
  else if(type === 'PaymentInfo') {
    return {
      type: types.PAYMENT_INFO,
      info
    };
  }
  else if(type === 'SubmitOrder') {
    return {
      type: types.SUBMIT_ORDER,
      info
    };
  }

}
