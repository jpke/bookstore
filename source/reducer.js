import * as types from './actionTypes';

export const initialState = {
   userInfo: {
     firstName: "",
     lastName: "",
     email: ""
   },
   deliveryAddress: {
     lineOne: "",
     lineTwo: "",
     city: "",
     state: "",
     zip: ""
   },
   useDeliveryAddressAsBilling: false,
   billingAddress: {
     lineOne: "",
     lineTwo: "",
     city: "",
     state: "",
     zip: ""
   },
   creditCardInfo: {
     nameOnCard: "",
     cardNumber: "",
     securityCode: ""
   },
   availableBooks: [
     {
       title: "The Evolution of Civilizations",
       author: "Carroll Quigley",
     },
     {
       title: "How the Mind Works",
       author: "Stephan Pinker",

     }
   ],
   orders: [
     {
       title: "The Lexus and the Olive Tree",
       author: "Thomas Friedman",
       date: new Date("Sat Mar 18 2017 16:38:08 GMT-0400 (EDT)")
     },
     {
       title: "Wit and Wisdom",
       author: "Toby Reynolds",
       date: new Date("Fri Mar 17 2017 16:40:30 GMT-0400 (EDT)")
     }
   ],
   bookToOrder: "",
   newBook: {
     title: "",
     author: ""
   }
 }

 export default function reducer(state = initialState, action) {
   //declare vars
   let newOrder, orders, books;

   switch(action.type) {

    case types.SELECT_NEW_BOOK:
      return {
        ...state,
      bookToOrder: action.newBook
      };
    case types.ADMIN_ADD_BOOK:
      books = state.availableBooks.concat(action.info);
      return {
        ...state,
        availableBooks: books
      };
    case types.USER_INFO:
      return {
        ...state,
        userInfo: action.info
      };
    case types.DELIVERY_AND_BILLING:
      return {
        ...state,
        deliveryAddress: action.info,
        billingAddress: action.info,
        useDeliveryAddressAsBilling: action.useDeliveryAddressAsBilling
      };
    case types.DELIVERY:
      return {
        ...state,
        deliveryAddress: action.info,
        useDeliveryAddressAsBilling: action.useDeliveryAddressAsBilling
      };
    case types.BILLING:
      return {
        ...state,
        billingAddress: action.info
      };
    case types.PAYMENT_INFO:
      return {
        ...state,
        creditCardInfo: action.info
      };
    case types.SUBMIT_ORDER:
      newOrder = Object.assign({}, action.info);
      newOrder.date = new Date();
      orders = state.orders.concat(newOrder);
      books = state.availableBooks.slice();
      books = books.filter(book => book.title !== newOrder.title)
      return {
        ...state,
        orders: orders,
        availableBooks: books
      };

    default:
     return state;
   }
 }
