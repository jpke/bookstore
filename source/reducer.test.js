import React from 'react';
import ReactNative from 'react-native';
import reducer from './reducer';
import {initialState} from './reducer';
import * as types from './actionTypes';
import * as actions from './actions';

it('should handle SELECT_NEW_BOOK', () => {
  let state = initialState;
  expect(reducer(state, actions.selectNewBook({
    title: "The Evolution of Civilizations",
    author: "Carroll Quigley",
  }))).toMatchSnapshot();
})

it('should handle ADMIN_ADD_BOOK', () => {
  let state = initialState;
  expect(reducer(state, actions.submitInfo(
    {
      title: "New Book",
      author: "Famous Author"
    }, 'AdminAddBook'
  ))).toMatchSnapshot();
})
