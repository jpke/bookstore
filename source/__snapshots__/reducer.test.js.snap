// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle ADMIN_ADD_BOOK 1`] = `
Object {
  "availableBooks": Array [
    Object {
      "author": "Carroll Quigley",
      "title": "The Evolution of Civilizations",
    },
    Object {
      "author": "Stephan Pinker",
      "title": "How the Mind Works",
    },
    Object {
      "author": "Famous Author",
      "title": "New Book",
    },
  ],
  "billingAddress": Object {
    "city": "",
    "lineOne": "",
    "lineTwo": "",
    "state": "",
    "zip": "",
  },
  "bookToOrder": "",
  "creditCardInfo": Object {
    "cardNumber": "",
    "nameOnCard": "",
    "securityCode": "",
  },
  "deliveryAddress": Object {
    "city": "",
    "lineOne": "",
    "lineTwo": "",
    "state": "",
    "zip": "",
  },
  "newBook": Object {
    "author": "",
    "title": "",
  },
  "orders": Array [
    Object {
      "author": "Thomas Friedman",
      "date": 2017-03-18T20:38:08.000Z,
      "title": "The Lexus and the Olive Tree",
    },
    Object {
      "author": "Toby Reynolds",
      "date": 2017-03-17T20:40:30.000Z,
      "title": "Wit and Wisdom",
    },
  ],
  "useDeliveryAddressAsBilling": false,
  "userInfo": Object {
    "email": "",
    "firstName": "",
    "lastName": "",
  },
}
`;

exports[`should handle BILLING 1`] = `
Object {
  "availableBooks": Array [
    Object {
      "author": "Carroll Quigley",
      "title": "The Evolution of Civilizations",
    },
    Object {
      "author": "Stephan Pinker",
      "title": "How the Mind Works",
    },
  ],
  "billingAddress": Object {
    "city": "_",
    "lineOne": "_",
    "lineTwo": "_",
    "state": "_",
    "zip": "_",
  },
  "bookToOrder": "",
  "creditCardInfo": Object {
    "cardNumber": "",
    "nameOnCard": "",
    "securityCode": "",
  },
  "deliveryAddress": Object {
    "city": "",
    "lineOne": "",
    "lineTwo": "",
    "state": "",
    "zip": "",
  },
  "newBook": Object {
    "author": "",
    "title": "",
  },
  "orders": Array [
    Object {
      "author": "Thomas Friedman",
      "date": 2017-03-18T20:38:08.000Z,
      "title": "The Lexus and the Olive Tree",
    },
    Object {
      "author": "Toby Reynolds",
      "date": 2017-03-17T20:40:30.000Z,
      "title": "Wit and Wisdom",
    },
  ],
  "useDeliveryAddressAsBilling": false,
  "userInfo": Object {
    "email": "",
    "firstName": "",
    "lastName": "",
  },
}
`;

exports[`should handle DELIVERY 1`] = `
Object {
  "availableBooks": Array [
    Object {
      "author": "Carroll Quigley",
      "title": "The Evolution of Civilizations",
    },
    Object {
      "author": "Stephan Pinker",
      "title": "How the Mind Works",
    },
  ],
  "billingAddress": Object {
    "city": "",
    "lineOne": "",
    "lineTwo": "",
    "state": "",
    "zip": "",
  },
  "bookToOrder": "",
  "creditCardInfo": Object {
    "cardNumber": "",
    "nameOnCard": "",
    "securityCode": "",
  },
  "deliveryAddress": Object {
    "city": "_",
    "lineOne": "_",
    "lineTwo": "_",
    "state": "_",
    "zip": "_",
  },
  "newBook": Object {
    "author": "",
    "title": "",
  },
  "orders": Array [
    Object {
      "author": "Thomas Friedman",
      "date": 2017-03-18T20:38:08.000Z,
      "title": "The Lexus and the Olive Tree",
    },
    Object {
      "author": "Toby Reynolds",
      "date": 2017-03-17T20:40:30.000Z,
      "title": "Wit and Wisdom",
    },
  ],
  "useDeliveryAddressAsBilling": false,
  "userInfo": Object {
    "email": "",
    "firstName": "",
    "lastName": "",
  },
}
`;

exports[`should handle DELIVERY_AND_BILLING 1`] = `
Object {
  "availableBooks": Array [
    Object {
      "author": "Carroll Quigley",
      "title": "The Evolution of Civilizations",
    },
    Object {
      "author": "Stephan Pinker",
      "title": "How the Mind Works",
    },
  ],
  "billingAddress": Object {
    "city": "_",
    "lineOne": "_",
    "lineTwo": "_",
    "state": "_",
    "zip": "_",
  },
  "bookToOrder": "",
  "creditCardInfo": Object {
    "cardNumber": "",
    "nameOnCard": "",
    "securityCode": "",
  },
  "deliveryAddress": Object {
    "city": "_",
    "lineOne": "_",
    "lineTwo": "_",
    "state": "_",
    "zip": "_",
  },
  "newBook": Object {
    "author": "",
    "title": "",
  },
  "orders": Array [
    Object {
      "author": "Thomas Friedman",
      "date": 2017-03-18T20:38:08.000Z,
      "title": "The Lexus and the Olive Tree",
    },
    Object {
      "author": "Toby Reynolds",
      "date": 2017-03-17T20:40:30.000Z,
      "title": "Wit and Wisdom",
    },
  ],
  "useDeliveryAddressAsBilling": true,
  "userInfo": Object {
    "email": "",
    "firstName": "",
    "lastName": "",
  },
}
`;

exports[`should handle PAYMENT_INFO 1`] = `
Object {
  "availableBooks": Array [
    Object {
      "author": "Carroll Quigley",
      "title": "The Evolution of Civilizations",
    },
    Object {
      "author": "Stephan Pinker",
      "title": "How the Mind Works",
    },
  ],
  "billingAddress": Object {
    "city": "",
    "lineOne": "",
    "lineTwo": "",
    "state": "",
    "zip": "",
  },
  "bookToOrder": "",
  "creditCardInfo": Object {
    "cardNumber": "_",
    "nameOnCard": "_",
    "securityCode": "_",
  },
  "deliveryAddress": Object {
    "city": "",
    "lineOne": "",
    "lineTwo": "",
    "state": "",
    "zip": "",
  },
  "newBook": Object {
    "author": "",
    "title": "",
  },
  "orders": Array [
    Object {
      "author": "Thomas Friedman",
      "date": 2017-03-18T20:38:08.000Z,
      "title": "The Lexus and the Olive Tree",
    },
    Object {
      "author": "Toby Reynolds",
      "date": 2017-03-17T20:40:30.000Z,
      "title": "Wit and Wisdom",
    },
  ],
  "useDeliveryAddressAsBilling": false,
  "userInfo": Object {
    "email": "",
    "firstName": "",
    "lastName": "",
  },
}
`;

exports[`should handle SELECT_NEW_BOOK 1`] = `
Object {
  "availableBooks": Array [
    Object {
      "author": "Carroll Quigley",
      "title": "The Evolution of Civilizations",
    },
    Object {
      "author": "Stephan Pinker",
      "title": "How the Mind Works",
    },
  ],
  "billingAddress": Object {
    "city": "",
    "lineOne": "",
    "lineTwo": "",
    "state": "",
    "zip": "",
  },
  "bookToOrder": Object {
    "author": "Carroll Quigley",
    "title": "The Evolution of Civilizations",
  },
  "creditCardInfo": Object {
    "cardNumber": "",
    "nameOnCard": "",
    "securityCode": "",
  },
  "deliveryAddress": Object {
    "city": "",
    "lineOne": "",
    "lineTwo": "",
    "state": "",
    "zip": "",
  },
  "newBook": Object {
    "author": "",
    "title": "",
  },
  "orders": Array [
    Object {
      "author": "Thomas Friedman",
      "date": 2017-03-18T20:38:08.000Z,
      "title": "The Lexus and the Olive Tree",
    },
    Object {
      "author": "Toby Reynolds",
      "date": 2017-03-17T20:40:30.000Z,
      "title": "Wit and Wisdom",
    },
  ],
  "useDeliveryAddressAsBilling": false,
  "userInfo": Object {
    "email": "",
    "firstName": "",
    "lastName": "",
  },
}
`;

exports[`should handle SUBMIT_ORDER 1`] = `
Object {
  "availableBooks": Array [
    Object {
      "author": "Stephan Pinker",
      "title": "How the Mind Works",
    },
  ],
  "billingAddress": Object {
    "city": "",
    "lineOne": "",
    "lineTwo": "",
    "state": "",
    "zip": "",
  },
  "bookToOrder": "",
  "creditCardInfo": Object {
    "cardNumber": "",
    "nameOnCard": "",
    "securityCode": "",
  },
  "deliveryAddress": Object {
    "city": "",
    "lineOne": "",
    "lineTwo": "",
    "state": "",
    "zip": "",
  },
  "newBook": Object {
    "author": "",
    "title": "",
  },
  "orders": Array [
    Object {
      "author": "Thomas Friedman",
      "date": 2017-03-18T20:38:08.000Z,
      "title": "The Lexus and the Olive Tree",
    },
    Object {
      "author": "Toby Reynolds",
      "date": 2017-03-17T20:40:30.000Z,
      "title": "Wit and Wisdom",
    },
    Object {
      "author": "Carroll Quigley",
      "date": 2017-03-18T20:38:08.000Z,
      "title": "The Evolution of Civilizations",
    },
  ],
  "useDeliveryAddressAsBilling": false,
  "userInfo": Object {
    "email": "",
    "firstName": "",
    "lastName": "",
  },
}
`;

exports[`should handle USER_INFO 1`] = `
Object {
  "availableBooks": Array [
    Object {
      "author": "Carroll Quigley",
      "title": "The Evolution of Civilizations",
    },
    Object {
      "author": "Stephan Pinker",
      "title": "How the Mind Works",
    },
  ],
  "billingAddress": Object {
    "city": "",
    "lineOne": "",
    "lineTwo": "",
    "state": "",
    "zip": "",
  },
  "bookToOrder": "",
  "creditCardInfo": Object {
    "cardNumber": "",
    "nameOnCard": "",
    "securityCode": "",
  },
  "deliveryAddress": Object {
    "city": "",
    "lineOne": "",
    "lineTwo": "",
    "state": "",
    "zip": "",
  },
  "newBook": Object {
    "author": "",
    "title": "",
  },
  "orders": Array [
    Object {
      "author": "Thomas Friedman",
      "date": 2017-03-18T20:38:08.000Z,
      "title": "The Lexus and the Olive Tree",
    },
    Object {
      "author": "Toby Reynolds",
      "date": 2017-03-17T20:40:30.000Z,
      "title": "Wit and Wisdom",
    },
  ],
  "useDeliveryAddressAsBilling": false,
  "userInfo": Object {
    "email": "_",
    "firstName": "_",
    "lastName": "_",
  },
}
`;
